core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/DedupeOperator.java

```diff
@@ -0,0 +1,160 @@
+import com.google.common.collect.ImmutableList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import lombok.EqualsAndHashCode;
+
+/**
+ * Dedupe operator. Dedupe the input {@link ExprValue} by using the {@link
+ * DedupeOperator#dedupeList} The result order follow the input order.
+ */
+@EqualsAndHashCode
+public class DedupeOperator extends PhysicalPlan {
+  private final PhysicalPlan input;
+  private final List<Expression> dedupeList;
+  private final Integer allowedDuplication;
+  private final Boolean keepEmpty;
+  private final Boolean consecutive;
+
+  @EqualsAndHashCode.Exclude private final Deduper<List<ExprValue>> deduper;
+  @EqualsAndHashCode.Exclude private ExprValue next;
+
+  private static final Integer ALL_ONE_DUPLICATION = 1;
+  private static final Boolean IGNORE_EMPTY = false;
+  private static final Boolean NON_CONSECUTIVE = false;
+
+  private static final Predicate<ExprValue> NULL_OR_MISSING = v -> v.isNull() || v.isMissing();
+
+  public DedupeOperator(PhysicalPlan input, List<Expression> dedupeList) {
+    this(input, dedupeList, ALL_ONE_DUPLICATION, IGNORE_EMPTY, NON_CONSECUTIVE);
+  }
+
+  public DedupeOperator(
+      PhysicalPlan input,
+      List<Expression> dedupeList,
+      Integer allowedDuplication,
+      Boolean keepEmpty,
+      Boolean consecutive) {
+    this.input = input;
+    this.dedupeList = dedupeList;
+    this.allowedDuplication = allowedDuplication;
+    this.keepEmpty = keepEmpty;
+    this.consecutive = consecutive;
+    this.deduper = this.consecutive ? new ConsecutiveDeduper<>() : new HistoricalDeduper<>();
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitDedupe(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    while (input.hasNext()) {
+      ExprValue next = input.next();
+      if (keep(next)) {
+        this.next = next;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public ExprValue next() {
+    return this.next;
+  }
+
+  /**
+   * Test the {@link ExprValue} should be keep or ignore
+   *
+   * <p>If any value evaluted by {@link DedupeOperator#dedupeList} is NULL or MISSING, then the *
+   * return value is decided by keepEmpty option, default value is ignore.
+   *
+   * @param value
+   * @return true: keep, false: ignore
+   */
+  public boolean keep(ExprValue value) {
+    BindingTuple bindingTuple = value.bindingTuples();
+    ImmutableList.Builder<ExprValue> dedupeKeyBuilder = new ImmutableList.Builder<>();
+    for (Expression expression : dedupeList) {
+      ExprValue exprValue = expression.valueOf(bindingTuple);
+      if (NULL_OR_MISSING.test(exprValue)) {
+        return keepEmpty;
```

Comment:
Probably safer to do null check before return in case this `Boolean` is null? Or verify this in constructor and save as `boolean` if it's required non-null?

Reply:
Done. add the @NonNull at the constructor.